libs/model-assessment/src/lib/ModelAssessmentDashboard/Controls/ModelOverview/ModelOverview.tsx

```diff
@@ -167,6 +175,26 @@ export class ModelOverview extends React.Component<
         };
       });
 
+    let chartCohorts: ErrorCohort[];
+    let someCohortSelected: boolean;
+    let selectedChartCohorts: number[];
+    let labeledStatistics: ILabeledStatistic[][];
+    if (this.state.datasetCohortChartIsVisible) {
+      chartCohorts = this.context.errorCohorts;
+      someCohortSelected =
+        this.state.selectedDatasetCohorts !== undefined &&
+        this.state.selectedDatasetCohorts.length > 0;
+      selectedChartCohorts = this.state.selectedDatasetCohorts ?? [];
+      labeledStatistics = datasetCohortLabeledStatistics;
+    } else {
+      chartCohorts = featureBasedCohorts;
+      someCohortSelected =
+        this.state.selectedFeatureBasedCohorts !== undefined &&
+        this.state.selectedFeatureBasedCohorts.length > 0;
+      selectedChartCohorts = this.state.selectedFeatureBasedCohorts ?? [];
+      labeledStatistics = featureBasedCohortLabeledStatistics;
+    }
```

Comment:
Seems a lot logic in the render.
This may slow the component rendering, if any of them is time consuming.
For long run process, it is better to save to state and use componentDidUpdate to check and update.

Reply:
Yes! There are a few tiny things that don't matter much performance-wise, but the big pieces are 
- generating the cohorts from the selected features (selected via dropdown)
- generating the metrics for those cohorts and the existing user-created dataset cohorts.

I don't anticipate this component re-rendering at all, though. You do an initial render when showing it the first time. When changing between features you necessarily have to recalculate the metrics, of course. There's no way to precalculate them because there are simply too many possible combinations. After that, I don't think it should be re-rendering with the same cohorts at all. Let me know if I'm missing something!